testdata/go/v3/monitoring/memcached-operator/monitoring/metricsdocs/metricsdocs.go

```diff
@@ -0,0 +1,119 @@
+/*
+Copyright 2022.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package main
+
+import (
+	"fmt"
+	"sort"
+
+	"github.com/example/memcached-operator/monitoring"
+)
+
+// constant parts of the file
+const (
+	title      = "# Memcached Operator Metrics\n"
+	background = "This document aims to help users that are not familiar with metrics exposed by the memcached-operator.\n" +
+		"All metrics documented here are auto-generated by the utility tool \"monitoring/metricsdocs\" and reflects exactly what is being exposed.\n\n"
+
+	KVSpecificMetrics = "## Memcached Operator Metrics List\n"
+
+	opening = title +
+		background +
+		KVSpecificMetrics
+
+	// footer
+	footerHeading = "## Developing new metrics\n"
+	footerContent = "After developing new metrics or changing old ones, please run \"make generate-metricsdocs\" to regenerate this document.\n\n" +
+		"If you feel that the new metric doesn't follow these rules, please change \"monitoring/metricsdocs\" according to your needs.\n"
+
+	footer = footerHeading + footerContent
+)
+
+// MemcachedMetricList contains the name, description, and type for each metric.
+func main() {
+	MemcachedMetricList := metricDescriptionListToMetricList(monitoring.ListMetrics())
+	sort.Sort(MemcachedMetricList)
+	writeToFile(MemcachedMetricList)
+}
+
+// writeToFile receives a list of metrics and writes them to a file.
+func writeToFile(metricsList metricList) {
+	fmt.Print(opening)
+	metricsList.writeOut()
+	fmt.Print(footer)
+}
+
+// metric is an exported struct that defines the metric
+// name, description, and type as a new type named metric.
+type metric struct {
+	name        string
+	description string
+	metricType  string
+}
+
+// metricDescriptionToMetric receives a metric of type Metric defined
+// in ./monitoring/metrics/util/util.go, and returns as a metric type.
+func metricDescriptionToMetric(md monitoring.MetricDescription) metric {
```

Comment:
These helpers which are generic (like listing, comparing etc) might be good candidates for `operator-lib`: https://github.com/operator-framework/operator-lib. Scaffolding them in every user's project might me unnecessary imo.

cc: @everettraven @jberkhahn @camilamacedo86

Reply:
What we could generate here is the `MemcachedMetricList`, the rest of the helpers are generic and can take in any metricDescriptors. These are not a part of reconciler code either and not related to registering metrics. This section is where, as @camilamacedo86 mentioned, we could have a library.

Reply:
> // metricDescriptionToMetric receives a metric of type Metric defined
> // in ./monitoring/metrics/util/util.go, and returns as a metric type.
Please ignore the comments lines (outdated and wrong).

I agree with using libs, you think we can implement this in a later pr?